File: backend/models/Post.js
import mongoose from 'mongoose';
import postSchema from '../schemas/db/postSchema.js';

export const Post = mongoose.model('Post', postSchema);

const postPopulate = (query) =>
  query
    .populate('user', ['username', 'isAdmin'])
    .populate('likes', ['username'])
    .populate('comments.user', ['username', 'isAdmin']);

export const createPost = (data) => Post.create(data);

export const countPosts = () => Post.countDocuments();

export const findPostsPaginated = ({ skip, limit }) =>
  postPopulate(
    Post.find(
      {},
      {
        title: 1,
        content: 1,
        user: 1,
        likes: 1,
        comments: 1,
        image: 1,
        createdAt: 1,
        updatedAt: 1,
      },
    )
      .sort({ createdAt: -1 })
      .skip(skip)
      .limit(limit),
  );

export const findPostById = (id) => postPopulate(Post.findById(id));

export const findPostLikesById = (id) => Post.findById(id).select('likes').lean();

export const updatePostById = (id, update, options) => Post.findByIdAndUpdate(id, update, options);

export const pushCommentToPost = async (postId, comment) => {
  const post = await Post.findById(postId);
  if (!post) return null;
  post.comments.push(comment);
  await post.save();
  return postPopulate(Post.findById(post._id));
};

File: backend-go/models/post.go
package models

import (
	"time"

	"go.mongodb.org/mongo-driver/v2/bson"
)

type Post struct {
	ID      bson.ObjectID   `bson:"_id,omitempty" json:"id"`
	Title   string          `bson:"title" json:"title"`
	Content string          `bson:"content" json:"content"`
	User    bson.ObjectID   `bson:"user" json:"user"`
	Likes   []bson.ObjectID `bson:"likes" json:"likes"`
	Image     Image     `bson:"image" json:"image"`
	Comments  []Comment `bson:"comments" json:"comments"`
	CreatedAt time.Time `bson:"createdAt" json:"createdAt"`
	UpdatedAt time.Time `bson:"updatedAt" json:"updatedAt"`
	Version   int       `bson:"__v" json:"v"`
}

type Comment struct {
	ID        bson.ObjectID `bson:"_id,omitempty" json:"id"`
	Text      string        `bson:"text" json:"text"`
	User      bson.ObjectID `bson:"user" json:"user"`
	CreatedAt time.Time     `bson:"createdAt" json:"createdAt"`
}

type Image struct {
	URL      string `bson:"url" json:"url"`
	PublicID string `bson:"public_id" json:"public_id"`
}
